cmd/conduit/root/root.go

```diff
@@ -0,0 +1,262 @@
+
+	// Map preview features
+	c.cfg.Preview.PipelineArchV2 = c.flags.PreviewPipelineArchV2
+
+	// Map development profiling
+	c.cfg.Dev.CPUProfile = c.flags.DevCPUProfile
+	c.cfg.Dev.MemProfile = c.flags.DevMemProfile
+	c.cfg.Dev.BlockProfile = c.flags.DevBlockProfile
+
+	// Update paths
+	c.cfg.DB.SQLite.Path = c.flags.DBSQLitePath
+	c.cfg.DB.Badger.Path = c.flags.DBBadgerPath
+	c.cfg.Pipelines.Path = c.flags.PipelinesPath
+	c.cfg.Connectors.Path = c.flags.ConnectorsPath
+	c.cfg.Processors.Path = c.flags.ProcessorsPath
+}
+
+func (c *RootCommand) updateFlagValuesFromConfig() {
+	// Map database configuration
+	c.flags.DBType = c.cfg.DB.Type
+	c.flags.DBPostgresConnectionString = c.cfg.DB.Postgres.ConnectionString
+	c.flags.DBPostgresTable = c.cfg.DB.Postgres.Table
+	c.flags.DBSQLiteTable = c.cfg.DB.SQLite.Table
+
+	// Map API configuration
+	c.flags.APIEnabled = c.cfg.API.Enabled
+	c.flags.APIHTTPAddress = c.cfg.API.HTTP.Address
+	c.flags.APIGRPCAddress = c.cfg.API.GRPC.Address
+
+	// Map logging configuration
+	c.flags.LogLevel = c.cfg.Log.Level
+	c.flags.LogFormat = c.cfg.Log.Format
+
+	// Map pipeline configuration
+	c.flags.PipelinesExitOnDegraded = c.cfg.Pipelines.ExitOnDegraded
+	c.flags.PipelinesErrorRecoveryMinDelay = c.cfg.Pipelines.ErrorRecovery.MinDelay
+	c.flags.PipelinesErrorRecoveryMaxDelay = c.cfg.Pipelines.ErrorRecovery.MaxDelay
+	c.flags.PipelinesErrorRecoveryBackoffFactor = c.cfg.Pipelines.ErrorRecovery.BackoffFactor
+	c.flags.PipelinesErrorRecoveryMaxRetries = c.cfg.Pipelines.ErrorRecovery.MaxRetries
+	c.flags.PipelinesErrorRecoveryMaxRetriesWindow = c.cfg.Pipelines.ErrorRecovery.MaxRetriesWindow
+
+	// Map schema registry configuration
+	c.flags.SchemaRegistryType = c.cfg.SchemaRegistry.Type
+	c.flags.SchemaRegistryConfluentConnectionString = c.cfg.SchemaRegistry.Confluent.ConnectionString
+
+	// Map preview features
+	c.flags.PreviewPipelineArchV2 = c.cfg.Preview.PipelineArchV2
+
+	// Map development profiling
+	c.flags.DevCPUProfile = c.cfg.Dev.CPUProfile
+	c.flags.DevMemProfile = c.cfg.Dev.MemProfile
+	c.flags.DevBlockProfile = c.cfg.Dev.BlockProfile
+
+	// Update paths
+	c.flags.DBSQLitePath = c.cfg.DB.SQLite.Path
+	c.flags.DBBadgerPath = c.cfg.DB.Badger.Path
+	c.flags.PipelinesPath = c.cfg.Pipelines.Path
+	c.flags.ConnectorsPath = c.cfg.Connectors.Path
+	c.flags.ProcessorsPath = c.cfg.Processors.Path
+}
+
+func (c *RootCommand) Execute(_ context.Context) error {
+	if c.flags.Version {
+		_, _ = fmt.Fprintf(os.Stdout, "%s\n", conduit.Version(true))
+		return nil
+	}
+
+	// 1. Load conduit configuration file and update general config.
+	if err := internal.LoadConfigFromFile(c.flags.ConduitConfigPath, &c.cfg); err != nil {
+		return err
+	}
+
+	// 2. Load environment variables and update general config.
+	if err := internal.LoadConfigFromEnv(&c.cfg); err != nil {
+		return err
+	}
+
+	// 3. Update the general config from flags.
+	c.updateConfigFromFlags()
```

Comment:
IIUC, these 3 steps can be done with a single Viper instance: https://gist.github.com/hariso/5ca078ba50b4916b4ca48673165dcc90, since Viper supports reading from multiple sources simultaneously. The precedence is the same as in Conduit: https://github.com/spf13/viper?tab=readme-ov-file#why-viper

Reply:
Trying it out (still need to fully test this) https://github.com/ConduitIO/conduit/pull/1983/commits/b6abfb6908e058f1244db3ee1d42e16e60e17afa